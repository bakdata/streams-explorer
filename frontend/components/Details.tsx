import { CopyOutlined } from "@ant-design/icons";
import { Alert, Button, Descriptions, message, Space, Spin } from "antd";
import copy from "copy-to-clipboard";
import React from "react";
import ReactJson from "react-json-view";
import {
  NodeInfoListItem,
  useGetLinkingApiNodeLinkingNodeIdGet,
  useGetNodeInfoApiNodeNodeIdGet,
} from "./api/fetchers";
import Schema from "./Schema";

interface DetailsProps {
  nodeId: string;
}

const Details = ({ nodeId }: DetailsProps) => {
  const { data, loading, error } = useGetNodeInfoApiNodeNodeIdGet({
    node_id: nodeId,
  });
  if (loading) {
    return (
      <div className="loadingSpinnerContainer">
        <Spin tip="Loading..." />
      </div>
    );
  }
  if (!data || error) {
    return (
      <Alert
        data-testid="no-node-info"
        message="No node details available"
        type="warning"
      />
    );
  }
  return (
    <Descriptions layout="horizontal" bordered column={1} size="small">
      <Descriptions.Item className="descItem" label="Type" key={1}>
        {data.node_type}
      </Descriptions.Item>
      {data.info.map((nodeInfoListItem) => (
        <Descriptions.Item
          className="descItem"
          label={nodeInfoListItem.name}
          key={nodeInfoListItem.name}
        >
          <NodeInfoDetail infoListItem={nodeInfoListItem} nodeId={nodeId} />
        </Descriptions.Item>
      ))}
    </Descriptions>
  );
};

interface NodeInfoDetailProps {
  infoListItem: NodeInfoListItem;
  nodeId: string;
}

const NodeInfoDetail = ({ infoListItem, nodeId }: NodeInfoDetailProps) => {
  switch (infoListItem.type) {
    case "json":
      return infoListItem.name === "Schema" ? (
        <Schema nodeId={nodeId} />
      ) : (
        <div className="jsonDetail">
          <ReactJson
            name={false}
            src={infoListItem.value as object}
            displayDataTypes={false}
            collapsed={false}
          />
        </div>
      );

    case "link":
      return <LinkInfo nodeId={nodeId} infoListItem={infoListItem} />;
    default:
      return <React.Fragment>{infoListItem.value}</React.Fragment>;
  }
};

const LinkInfo = ({ infoListItem, nodeId }: NodeInfoDetailProps) => {
  let linkType =
    infoListItem.value !== "" && typeof infoListItem.value == "string"
      ? infoListItem.value
      : undefined;
  const { data: linkToService, loading } = useGetLinkingApiNodeLinkingNodeIdGet(
    {
      node_id: nodeId,
      queryParams: { link_type: linkType },
    }
  );

  if (loading) {
    return <Spin tip="Loading link..." />;
  }
  if (linkToService) {
    return (
      <Space>
        <Button href={linkToService} type="primary" target="_blank">
          {infoListItem.name}
        </Button>

        <Button
          icon={<CopyOutlined />}
          onClick={() => {
            message.success("Copied link to clipboard");
            copy(linkToService);
          }}
        />
      </Space>
    );
  }
  return <>Could not get link to service</>;
};

export default Details;
