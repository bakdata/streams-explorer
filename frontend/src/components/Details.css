.Info {
  padding-left: 14px;
}

.descItem {
  vertical-align: top;
}

.ant-descriptions-item-label {
  width: 16em;
}

.details {
  width: 100%;
  text-align: start;
}

.loadingSpinnerContainer {
  width: 100%;
  text-align: center;
}

.jsonDetail {
  font-size: 11px;
}
