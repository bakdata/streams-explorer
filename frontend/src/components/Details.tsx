import React from "react";
import {
  useNodeInfoApiNodeNodeIdGet,
  NodeInfoListItem,
  useLinkingApiNodeLinkingNodeIdGet,
} from "../api/fetchers";
import { Spin, Alert, Descriptions, Button, Space, message } from "antd";
import ReactJson from "react-json-view";
import { CopyOutlined } from "@ant-design/icons";
import copy from "copy-to-clipboard";

interface DetailsProps {
  nodeID: string;
}

const Details = ({ nodeID }: DetailsProps) => {
  const { data, loading, error } = useNodeInfoApiNodeNodeIdGet({
    node_id: nodeID,
  });
  if (loading) {
    return (
      <div className="loadingSpinnerContainer">
        <Spin tip="Loading..." />
      </div>
    );
  }
  if (!data || error) {
    return (
      <Alert
        data-testid="no-node-info"
        message="No node details available"
        type="warning"
      />
    );
  }
  return (
    <Descriptions layout="horizontal" bordered column={1} size="small">
      <Descriptions.Item className="descItem" label="Type" key={1}>
        {data.node_type}
      </Descriptions.Item>
      {data.info.map((nodeInfoListItem) => (
        <Descriptions.Item
          className="descItem"
          label={nodeInfoListItem.name}
          key={nodeInfoListItem.name}
        >
          <NodeInfoDetail infoListItem={nodeInfoListItem} nodeID={nodeID} />
        </Descriptions.Item>
      ))}
    </Descriptions>
  );
};

interface NodeInfoDetailProps {
  infoListItem: NodeInfoListItem;
  nodeID: string;
}

const NodeInfoDetail = ({ infoListItem, nodeID }: NodeInfoDetailProps) => {
  switch (infoListItem.type) {
    case "json":
      return (
        <div className="jsonDetail">
          <ReactJson
            src={infoListItem.value}
            displayDataTypes={false}
            collapsed={false}
          />
        </div>
      );

    case "link":
      return <LinkInfo nodeID={nodeID} infoListItem={infoListItem} />;
    default:
      return <React.Fragment>{infoListItem.value}</React.Fragment>;
  }
};

const LinkInfo = ({ infoListItem, nodeID }: NodeInfoDetailProps) => {
  let linkType =
    infoListItem.value !== "" && typeof infoListItem.value == "string"
      ? infoListItem.value
      : undefined;
  const { data: linkToService, loading } = useLinkingApiNodeLinkingNodeIdGet({
    node_id: nodeID,
    queryParams: { link_type: linkType },
  });

  if (loading) {
    return <Spin tip="Loading link..." />;
  }
  if (linkToService) {
    return (
      <Space>
        <Button href={linkToService} type="primary" target="_blank">
          {infoListItem.name}
        </Button>

        <Button
          icon={<CopyOutlined />}
          onClick={() => {
            message.success("Copied link to clipboard");
            copy(linkToService);
          }}
        />
      </Space>
    );
  }
  return <>Could not get link to service</>;
};

export default Details;
