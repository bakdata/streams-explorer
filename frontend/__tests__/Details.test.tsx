import nock from "nock";
import React from "react";
import { RestfulProvider } from "restful-react";

import {
  fireEvent,
  render,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import Details from "../components/Details";

describe("display node information", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  it("should handle node not found error", async () => {
    nock("http://localhost").get("/api/node/1").reply(404, {
      detail: 'Could not find information for node with name "fake-node"',
    });

    const { findByTestId, asFragment } = render(
      <RestfulProvider base="http://localhost">
        <Details nodeId="1" />
      </RestfulProvider>
    );

    await findByTestId("no-node-info");
    expect(asFragment()).toMatchSnapshot();
  });

  it("should handle connector info", async () => {
    nock("http://localhost").get("/api/node/2").reply(200, {
      node_id: "demo-ames-sink",
      node_type: "connector",
      info: [],
    });

    const { findByText, asFragment } = render(
      <RestfulProvider base="http://localhost">
        <Details nodeId="2" />
      </RestfulProvider>
    );

    await findByText("connector");
    expect(asFragment()).toMatchSnapshot();
  });

  it("should handle connector info", async () => {
    nock("http://localhost").get("/api/node/2").reply(200, {
      node_id: "demo-ames-sink",
      node_type: "connector",
      info: [],
    });

    const { findByText, asFragment } = render(
      <RestfulProvider base="http://localhost">
        <Details nodeId="2" />
      </RestfulProvider>
    );

    await findByText("connector");
    expect(asFragment()).toMatchSnapshot();
  });

  it("should handle streams app info", async () => {
    nock("http://localhost")
      .get("/api/node/atm-fraud-transactionavroproducer")
      .reply(200, {
        node_id: "atm-fraud-transactionavroproducer",
        node_type: "streaming-app",
        info: [
          { name: "Kibana Logs", value: "kibana", type: "link" },
          {
            name: "Labels",
            value: {
              app: "atm-fraud-transactionavroproducer",
              "app.kubernetes.io/managed-by": "Helm",
              chart: "streams-app-0.1.0",
              pipeline: "streams-explorer",
              release: "demo-transactionavroproducer",
            },
            type: "json",
          },
        ],
      });

    const nockLinking = nock("http://localhost")
      .get("/api/node/linking/atm-fraud-transactionavroproducer")
      .query({ link_type: "kibana" })
      .reply(
        200,
        "http://localhost:5601/app/kibana#/discover?_a=(columns:!(_source),query:(language:lucene,query:'kubernetes.labels.app:%20%22atm-fraud-transactionavroproducer%22'))"
      );
    const { findByText, asFragment, queryByText } = render(
      <RestfulProvider base="http://localhost">
        <Details nodeId="atm-fraud-transactionavroproducer" />
      </RestfulProvider>
    );

    await findByText("streaming-app");
    await waitForElementToBeRemoved(() => queryByText("Loading link..."));
    expect(nockLinking.isDone()).toBeTruthy();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should handle topic info", async () => {
    nock("http://localhost")
      .get("/api/node/atm-fraud-incoming-transactions-topic")
      .reply(200, {
        node_id: "atm-fraud-incoming-transactions-topic",
        node_type: "topic",
        info: [
          { name: "Topic Monitoring", value: "grafana", type: "link" },
          {
            name: "Schema",
            value: {},
            type: "json",
          },
        ],
      });

    nock("http://localhost")
      .get("/api/node/atm-fraud-incoming-transactions-topic/schema")
      .reply(200, [1, 2]);

    const nockSchema2 = nock("http://localhost")
      .get("/api/node/atm-fraud-incoming-transactions-topic/schema/2")
      .reply(200, {
        type: "record",
        name: "Transaction2",
        namespace: "com.bakdata.kafka",
        fields: [
          {
            name: "transaction_id",
            type: { type: "string", "avro.java.string": "String" },
          },
          {
            name: "account_id",
            type: { type: "string", "avro.java.string": "String" },
          },
          { name: "amount", type: "int" },
          {
            name: "atm",
            type: { type: "string", "avro.java.string": "String" },
            default: "",
          },
          {
            name: "timestamp",
            type: { type: "long", logicalType: "timestamp-millis" },
          },
          {
            name: "location",
            type: {
              type: "record",
              name: "Location",
              fields: [
                { name: "latitude", type: "double" },
                { name: "longitude", type: "double" },
              ],
            },
          },
        ],
      });

    const nockSchema1 = nock("http://localhost")
      .get("/api/node/atm-fraud-incoming-transactions-topic/schema/1")
      .reply(200, {
        type: "record",
        name: "Transaction1",
        namespace: "com.bakdata.kafka",
        fields: [
          {
            name: "transaction_id",
            type: { type: "string", "avro.java.string": "String" },
          },
        ],
      });

    nock("http://localhost")
      .get(
        "/api/node/linking/atm-fraud-incoming-transactions-topic?link_type=grafana"
      )
      .reply(
        200,
        "http://localhost:3000/d/path/to/dashboard?var-topics=atm-fraud-incoming-transactions-topic"
      );

    const { getByText, findByText, getByTestId } = render(
      <RestfulProvider base="http://localhost">
        <Details nodeId="atm-fraud-incoming-transactions-topic" />
      </RestfulProvider>
    );

    await findByText("v2"); // get dropdown menu for schema version
    let schemaVersion = getByText("v2");
    expect(nockSchema2.isDone()).toBeTruthy();
    expect(nockSchema1.isDone()).toBeFalsy();
    const schema2 = getByTestId("schema");
    expect(schema2).toMatchSnapshot();

    fireEvent.mouseOver(getByTestId("schema-version"));
    await waitFor(() => {
      const menu = getByTestId("schema-version-select");
      fireEvent.mouseOver(menu);
      const v1 = getByText("v1");
      fireEvent.click(v1);
    });

    await waitFor(() => {
      expect(schemaVersion).toHaveTextContent("v1");
      expect(nockSchema1.isDone()).toBeTruthy();
    });

    await waitFor(() => {
      expect(schema2).not.toBeInTheDocument();
      const schema1 = getByTestId("schema");
      expect(schema1).toMatchSnapshot();
    });
  });

  it("should show error if schema versions are unavailable", async () => {
    nock("http://localhost")
      .get("/api/node/atm-fraud-incoming-transactions-topic")
      .reply(200, {
        node_id: "atm-fraud-incoming-transactions-topic",
        node_type: "topic",
        info: [
          {
            name: "Schema",
            value: {},
            type: "json",
          },
        ],
      });

    nock("http://localhost")
      .get("/api/node/atm-fraud-incoming-transactions-topic/schema")
      .reply(404);

    const { findByTestId } = render(
      <RestfulProvider base="http://localhost">
        <Details nodeId="atm-fraud-incoming-transactions-topic" />
      </RestfulProvider>
    );

    await findByTestId("no-schema-versions");
  });

  it("should show error if schema versions is empty", async () => {
    nock("http://localhost")
      .get("/api/node/atm-fraud-incoming-transactions-topic")
      .reply(200, {
        node_id: "atm-fraud-incoming-transactions-topic",
        node_type: "topic",
        info: [
          {
            name: "Schema",
            value: {},
            type: "json",
          },
        ],
      });

    nock("http://localhost")
      .get("/api/node/atm-fraud-incoming-transactions-topic/schema")
      .reply(200, []);

    const { findByTestId } = render(
      <RestfulProvider base="http://localhost">
        <Details nodeId="atm-fraud-incoming-transactions-topic" />
      </RestfulProvider>
    );

    await findByTestId("no-schema-versions");
  });

  it("should show error if schema is unavailable", async () => {
    nock("http://localhost")
      .get("/api/node/atm-fraud-incoming-transactions-topic")
      .reply(200, {
        node_id: "atm-fraud-incoming-transactions-topic",
        node_type: "topic",
        info: [
          {
            name: "Schema",
            value: {},
            type: "json",
          },
        ],
      });

    nock("http://localhost")
      .get("/api/node/atm-fraud-incoming-transactions-topic/schema")
      .reply(200, [1]);

    nock("http://localhost")
      .get("/api/node/atm-fraud-incoming-transactions-topic/schema/1")
      .reply(404);

    const { findByTestId } = render(
      <RestfulProvider base="http://localhost">
        <Details nodeId="atm-fraud-incoming-transactions-topic" />
      </RestfulProvider>
    );

    await findByTestId("no-schema");
  });
});
