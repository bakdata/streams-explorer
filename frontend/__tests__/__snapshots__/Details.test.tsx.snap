// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display node information should handle node not found error 1`] = `
<DocumentFragment>
  <div
    class="ant-alert ant-alert-warning ant-alert-no-icon"
    data-show="true"
    data-testid="no-node-info"
    role="alert"
  >
    <div
      class="ant-alert-content"
    >
      <div
        class="ant-alert-message"
      >
        No node details available
      </div>
      <div
        class="ant-alert-description"
      />
    </div>
  </div>
</DocumentFragment>
`;
